devito/petsc/iet/routines.py

```diff
@@ -393,26 +430,80 @@ def runsolve(self, solver_objs, objs, rhs_callback, injectsolve):
             dmda, solver_objs['x_global'], 'INSERT_VALUES', solver_objs['x_local']]
         )
 
-        calls = (rhs_call,
-                 dm_local_to_global_x,
-                 dm_local_to_global_b,
-                 snes_solve,
-                 dm_global_to_local_x)
-
-        return calls
+        return (
+            rhs_call,
+            local_x
+        ) + vec_replace_array + (
+            dm_local_to_global_x,
+            dm_local_to_global_b,
+            snes_solve,
+            dm_global_to_local_x,
+            BlankLine,
+        )
 
 
 def build_petsc_struct(iet, name, liveness):
     # Place all context data required by the shell routines into a struct
     # TODO: Clean this search up
     basics = FindSymbols('basics').visit(iet)
+    time_dims = [i for i in FindSymbols('dimensions').visit(iet)
+                 if isinstance(i, (TimeDimension, ModuloDimension))]
     avoid0 = [i for i in FindSymbols('indexedbases').visit(iet)
               if isinstance(i.function, PETScArray)]
-    avoid1 = FindSymbols('dimensions|writes').visit(iet)
+    avoid1 = [i for i in FindSymbols('dimensions|writes').visit(iet)
+              if i not in time_dims]
     fields = [data.function for data in basics if data not in avoid0+avoid1]
     return petsc_struct(name, fields, liveness)
 
 
+def time_dep_replace(injectsolve, target, solver_objs, objs, sregistry):
+    target_time = injectsolve.expr.lhs
+    target_time = [i for i, d in zip(target_time.indices,
+                                     target_time.dimensions) if d.is_Time]
+    assert len(target_time) == 1
+    target_time = target_time.pop()
+
+    ctype_str = ctypes_to_cstr(dtype_to_ctype(target.dtype))
+
+    class BarCast(Cast):
+        _base_typ = ctype_str
+
+    class StartPtr(LocalObject):
```

Comment:
could you have used this instead? 

https://github.com/devitocodes/devito/blob/master/devito/symbolics/extended_sympy.py#L830